api/provider/aws/job_runner.go

```diff
@@ -1,25 +1,256 @@
+	}
+
+	loadBalancer, err := r.loadBalancerProvider.Create(req)
+	if err != nil {
+		return err
+	}
+
+	return r.jobStore.SetJobResult(jobID, loadBalancer.LoadBalancerID)
+}
+
+func (r *JobRunner) createService(jobID, request string) error {
+	var req models.CreateServiceRequest
+	if err := json.Unmarshal([]byte(request), &req); err != nil {
+		return errors.New(errors.InvalidRequest, err)
+	}
+
+	service, err := r.serviceProvider.Create(req)
+	if err != nil {
+		return err
+	}
+
+	return r.jobStore.SetJobResult(jobID, service.ServiceID)
+}
+
+func (r *JobRunner) createTask(jobID, request string) error {
+	var req models.CreateTaskRequest
+	if err := json.Unmarshal([]byte(request), &req); err != nil {
+		return errors.New(errors.InvalidRequest, err)
+	}
+
+	task, err := r.taskProvider.Create(req)
+	if err != nil {
+		return err
+	}
+
+	return r.jobStore.SetJobResult(jobID, task.TaskID)
+}
+
+func (r *JobRunner) deleteDeploy(jobID, deployID string) error {
+	return r.deployProvider.Delete(deployID)
+}
+
+func (r *JobRunner) deleteEnvironment(jobID, environmentID string) error {
+	loadBalancers, err := r.loadBalancerProvider.List()
+	if err != nil {
+		return err
+	}
+
+	for _, loadBalancer := range loadBalancers {
+		if loadBalancer.EnvironmentID == environmentID {
+			if err := r.deleteLoadBalancer(jobID, loadBalancer.LoadBalancerID); err != nil {
+				return err
+			}
+		}
+	}
+
+	services, err := r.serviceProvider.List()
+	if err != nil {
+		return err
+	}
+
+	for _, service := range services {
+		if service.EnvironmentID == environmentID {
+			if err := r.deleteService(jobID, service.ServiceID); err != nil {
+				return err
+			}
+		}
+	}
+
+	tasks, err := r.taskProvider.List()
+	if err != nil {
+		return err
+	}
+
+	for _, task := range tasks {
+		if task.EnvironmentID == environmentID {
+			if err := r.deleteTask(jobID, task.TaskID); err != nil {
+				return err
+			}
 		}
+	}
 
-		return nil
+	return catchAndRetry(15, func() (shouldRetry bool, err error) {
+		if err := r.environmentProvider.Delete(environmentID); err != nil {
+			switch err := err.(type) {
+			case awserr.Error:
+				return err.Code() == "DependencyViolation", err
+			default:
+				return false, err
+			}
+		}
+
+		return false, nil
 	})
 }
 
-func deleteEnvironmentRunner(store job.Store, j models.Job) error {
-	return nil
+func (r *JobRunner) deleteLoadBalancer(jobID, loadBalancerID string) error {
+	return r.loadBalancerProvider.Delete(loadBalancerID)
+}
+
+func (r *JobRunner) deleteService(jobID, serviceID string) error {
+	return r.serviceProvider.Delete(serviceID)
+}
+
+func (r *JobRunner) deleteTask(jobID, taskID string) error {
+	return r.taskProvider.Delete(taskID)
+}
+
+func (r *JobRunner) updateEnvironment(jobID, request string) error {
+	var req models.UpdateEnvironmentRequest
+	if err := json.Unmarshal([]byte(request), &req); err != nil {
+		return errors.New(errors.InvalidRequest, err)
+	}
+
+	return r.environmentProvider.Update(req)
```

Comment:
Same things as the deletes - why don't we update the JobStore here?

Reply:
answered above